File: python/fourdgs/fourdgs/keyframe_delta_writer.py
# Copyright 2026 Avala AI
# SPDX-License-Identifier: Apache-2.0

"""Encoding a `keyframe-delta` file.

The caller supplies a **sequence of samples**: a population, with identities, stated at a
sequence of times. The encoder picks keyframes by cadence, quantizes every sample on one
set of grids, and writes each non-keyframe sample as the *difference of bins* against the
chunk it references.

Quantizing every sample up front, on grids derived from the whole sequence, is the part
that makes the model's error claim true. A delta is then an integer subtraction between
two bins on the same grid, the composition telescopes, and the reconstructed bin at any
depth is exactly the bin an absolute statement of that instant would have carried. An
encoder that instead quantized `x_j - x_{j-1}` would produce a file whose declared bounds
mean nothing after the second delta.

What this encoder does not do is decide *what* to update. Which gaussians moved enough to
be worth a byte, and how often to spend a keyframe, is rate control — encoder policy, and
the place encoders differentiate. This one updates a gaussian when any of its bins
changed, which is the simplest rule that is also correct.
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

from . import opcode as op
from . import records as rec
from .exceptions import InvalidInput
from .keyframe_delta import ABSOLUTE_IN_UPDATE, GOP_INVARIANT
from .model import GaussianSet
from .serialization import encode_stream


@dataclass
class Sample:
    """One population, at one instant, with identity.

    `ids` is aligned with the gaussians and is what a delta names them by. It is required
    rather than derived: the whole model rests on correspondence between samples, and a
    correspondence the encoder invented from row order is one the caller never asserted.
    """

    t0: float
    ids: np.ndarray
    gaussians: GaussianSet


@dataclass
class KeyframeDeltaOptions:
    """Cadence and mode. Everything else comes from the ordinary `WriteOptions`."""

    #: Samples per group of pictures. 1 writes every sample as a keyframe, which is legal
    #: and is the shape the registry's `frame-sequence` reservation describes.
    keyframe_every: int = 8
    #: `DELTA_MODE_CHAINED` references the previous chunk; `DELTA_MODE_KEYFRAME`
    #: references the group's keyframe. Chained is the default because it is smaller and
    #: its chain is contiguous in the file, so a range reader coalesces it into one
    #: request — see the design's section on why the choice is about cost and not error.
    delta_mode: int = rec.DELTA_MODE_CHAINED
    #: Sample indices to force a keyframe at, beyond the cadence. A producer that knows
    #: where a cut is puts one here so that instant costs two records however deep into a
    #: group it would otherwise have fallen.
    keyframe_at: tuple[int, ...] = field(default_factory=tuple)


def check_invariants(previous: dict[int, np.ndarray], current: dict[int, np.ndarray], rows, ids) -> None:
    """Refuse a sequence that changes a per-gaussian grid mid-group.

    `sigma_t`, `flags` and `window_index` derive the grids for velocity and birth time, so
    a bin difference across a change in any of them subtracts bins that live on different
    grids. The result is not an approximation, it is a number with no meaning — and it
    decodes silently into a wrong velocity rather than into an error, which is why this is
    the encoder's job to catch rather than the decoder's to survive.

    The fix is always available to the caller: a death and a birth, or a keyframe.
    """
    for attribute in sorted(GOP_INVARIANT):
        if attribute not in previous or attribute not in current:
            continue
        before = previous[attribute][rows]
        after = current[attribute]
        if not np.array_equal(before, after):
            row = int(np.argmax((before != after).any(axis=1) if before.ndim > 1 else (before != after)))
            raise InvalidInput(
                f"gaussian id {int(ids[row])} changes attribute {attribute} between samples, which is fixed "
                f"for a gaussian's lifetime within a group: the per-gaussian grids for velocity and birth "
                f"time are derived from it. Emit a keyframe, or a death and a birth.",
                code="invariant-changed-in-update",
            )


def delta_groups(
    reference_ids: np.ndarray,
    reference_bins: dict[int, np.ndarray],
    ids: np.ndarray,
    bins: dict[int, np.ndarray],
) -> tuple[np.ndarray, dict[int, np.ndarray], np.ndarray, dict[int, np.ndarray], np.ndarray]:
    """Split one sample against its reference into updates, births and deaths."""
    live = np.isin(ids, reference_ids)
    death_ids = np.setdiff1d(reference_ids, ids)
    birth_ids = ids[~live]
    birth_bins = {a: v[~live] for a, v in bins.items()}

    common_ids = ids[live]
    order = np.argsort(reference_ids, kind="stable")
    rows = order[np.searchsorted(reference_ids[order], common_ids)]

    common = {a: v[live] for a, v in bins.items()}
    check_invariants(reference_bins, common, rows, common_ids)

    # A gaussian is updated when any of its bins moved. Untouched means no bytes, which is
    # the property the whole model exists to buy.
    changed = np.zeros(common_ids.shape[0], dtype=bool)
    for attribute, values in common.items():
        if attribute in GOP_INVARIANT:
            continue
        before = reference_bins[attribute][rows]
        differs = values != before
        changed |= differs.any(axis=1) if differs.ndim > 1 else differs

    update_ids = common_ids[changed]
    update_rows = rows[changed]
    update_bins: dict[int, np.ndarray] = {}
    for attribute, values in common.items():
        if attribute in GOP_INVARIANT:
            continue
        touched = values[changed]
        if attribute in ABSOLUTE_IN_UPDATE:
            # Restated outright: the smallest-three basis changes whenever the largest
            # quaternion component does, so the three stored bins mean different
            # components either side of it and a difference would be nonsense.
            update_bins[attribute] = touched
        else:
            update_bins[attribute] = touched - reference_bins[attribute][update_rows]

    return update_ids, update_bins, birth_ids, birth_bins, death_ids


def encode_delta_streams(ids: np.ndarray, bins: dict[int, np.ndarray], *, codec: int, level: int) -> bytes:
    """One group's streams: the ids it names, then the attributes it carries."""
    if ids.size == 0:
        return b""
    out = [encode_stream(op.A_GAUSSIAN_ID, ids.reshape(-1, 1), codec=codec, level=level)]
    for attribute in sorted(bins):
        values = bins[attribute]
        values = values.reshape(-1, 1) if values.ndim == 1 else values
        out.append(encode_stream(attribute, values, channels=values.shape[1], codec=codec, level=level))
    return b"".join(out)


def death_streams(ids: np.ndarray, *, codec: int, level: int) -> bytes:
    if ids.size == 0:
        return b""
    return encode_stream(op.A_GAUSSIAN_ID, ids.reshape(-1, 1), codec=codec, level=level)


def is_keyframe(index: int, kd: KeyframeDeltaOptions) -> bool:
    return index == 0 or index in kd.keyframe_at or (kd.keyframe_every > 0 and index % kd.keyframe_every == 0)

File: swift/Sources/FourDGS/KeyframeDeltaWriter.swift
// Copyright 2026 Avala AI
// SPDX-License-Identifier: Apache-2.0

/// Authoring a `keyframe-delta` file: a sequence of populations in, bytes out.
///
/// ``SceneWriter`` authors the other temporal model — one population whose gaussians each
/// carry their own birth time. It cannot say the thing this model is for: the *same*
/// population, with identity, restated at a sequence of instants, with gaussians entering and
/// leaving (spec §11). So this is a separate writer rather than an option on that one, for
/// the same reason a Delta Chunk is its own record and not a flag on Chunk (spec §5.18).
///
/// Like ``SceneWriter``, this is a binding and not a second encoder: the sequence crosses into
/// the Rust core through ``CoreSeam`` and the bytes it returns are the bytes the reference
/// encoder writes for the same input. That is not merely convenient. A delta is a **difference
/// of bins, never a quantization of a difference** (spec §11.7) — which is what makes chained
/// deltas accumulate no error at any depth — and it holds only if every sample was quantized
/// up front on one set of grids derived from the whole sequence. An encoder that subtracted in
/// Swift and quantized afterwards would produce a file whose declared bounds mean nothing
/// after the second delta.

/// Which chunk a delta is stated against (spec §11.4).
public enum DeltaMode: UInt8, Sendable, Equatable {
    /// The reference is the keyframe at the head of the group. Always two records to
    /// reconstruct, however deep into the group the delta falls — the shape to place at a
    /// likely seek target such as a chapter boundary or a loop start.
    case keyframeReferenced = 0
    /// The reference is the state chunk immediately before this one. Smaller, and contiguous
    /// in the file so a range reader coalesces a chain into one request. The recommended
    /// default: chaining costs nothing in error here, so the choice is about cost alone.
    case chained = 1
}

/// One population, at one instant, with identity.
public struct KeyframeDeltaSample: Sendable, Equatable {
    /// When this state was stated. It is also the start of the interval the sample covers:
    /// sample `i` is valid over `[t0_i, t0_{i+1})`, and the last one runs to the duration.
    public var t0: Double
    /// One `gaussian_id` per gaussian, aligned with ``gaussians`` (spec §11.2).
    ///
    /// Required rather than derived from row order, because the whole model rests on
    /// correspondence between samples and a correspondence invented from row order is one the
    /// caller never asserted. Ids need not be dense, ordered, or start at zero; an id MUST NOT
    /// be reused after the gaussian carrying it dies.
    public var ids: [UInt32]
    public var gaussians: GaussianState

    public init(t0: Double, ids: [UInt32], gaussians: GaussianState) {
        self.t0 = t0
        self.ids = ids
        self.gaussians = gaussians
    }
}

/// Cadence and mode. The defaults are the reference encoder's own.
public struct KeyframeDeltaWriteOptions: Sendable, Equatable {
    /// Samples per group of pictures. `1` writes every sample as a keyframe, which is legal
    /// and is the shape the registry's `frame-sequence` reservation describes; `0` disables
    /// the cadence, leaving sample 0 and ``keyframeAt`` as the only keyframes.
    public var keyframeEvery: UInt32 = 8
    /// Which chunk each delta references. Per-sequence here; the format allows it per chunk.
    public var deltaMode: DeltaMode = .chained
    /// Sample indices to force a keyframe at, beyond the cadence. A producer that knows where
    /// a cut is puts one here so that instant is a whole restatement however deep into a group
    /// it would otherwise have fallen.
    public var keyframeAt: [UInt32] = []
    /// The error bounds the whole sequence is quantized against, and the Header's `profile`:
    /// `"fine"`, `"default"` or `"coarse"`. Unlike ``WriteOptions/profile`` — a free-form
    /// promise separate from the bounds — these are one value under this model, because the
    /// grids every sample is quantized on come from it.
    public var profile: String = "default"
    /// The Header's marginal visibility threshold, with the meaning it has on the other
    /// writer: it sets the support constant the per-gaussian velocity grid is derived from.
    public var cutoff: Double = 0.05
    /// The Header's `library`. Empty leaves the core's own default in place, so a caller that
    /// does not set it carries the same string every other default-configured writer does.
    public var library: String = ""
    /// The stream codec applied to every chunk's block. `0` is deflate, the format's default.
    public var codec: UInt8 = 0
    /// The codec's level.
    public var compressionLevel: UInt32 = 6

    public init() {}
}

/// Encode a sequence of populations into `keyframe-delta` bytes.
public enum KeyframeDeltaWriter {

    /// Encode `samples` over a `durationSec`-long timeline into `.4dgs` bytes.
    ///
    /// The samples must tile the timeline (spec §11.1): they are given in time order, the
    /// first starts at 0, each one's interval ends where the next begins, and the last ends at
    /// `durationSec`. The encoder derives each `t1` from the next sample's `t0` rather than
    /// checking the endpoints, so a sequence that misses this produces a file the indexed read
    /// path refuses as `non-tiling-chunks` while the streamed path accepts it — check the
    /// timeline you hand it, or read the result back on both paths as the tests here do.
    ///
    /// Throws the same typed errors ``SceneWriter`` does when the input is rejected: an empty
    /// sequence, a sample whose ids and gaussians are different lengths, a non-finite
    /// `sigmaT`, or a gaussian whose `sigmaT` or validity window changes inside a group. That
    /// last one is refused rather than written because those values *derive the grid* a bin
    /// difference is taken on (spec §11.5) — a file carrying such a change decodes silently
    /// into a wrong velocity rather than into an error, so it is the encoder's job to catch.
    /// The fix is always available: emit a death and a birth, or a keyframe.
    ///
    /// Spherical harmonics are not carried; a file written here declares `sh_degree` 0.
    public static func encode(
        _ samples: [KeyframeDeltaSample], durationSec: Double,
        options: KeyframeDeltaWriteOptions = KeyframeDeltaWriteOptions()
    ) throws -> [UInt8] {
        try Core.encodeKeyframeDelta(samples, durationSec: durationSec, options: options)
    }
}
